src/java/htsjdk/samtools/SAMRecord.java

```diff
@@ -514,27 +478,79 @@ else if (null != mHeader) {
      * @param mateReferenceIndex Must either equal NO_ALIGNMENT_REFERENCE_INDEX (-1) indicating no reference, or the
      *                       record must have a SAMFileHeader and the index must exist in the associated sequence
      *                       dictionary.
-     * @throws IllegalStateException if the SAMFileHeader is null for this record or the mate reference index is not
-     *         found in the sequence dictionary for this record.
+     * @throws IllegalStateException if the SAMFileHeader is null for this record
+     * @throws IllegalArgumentException if the mate reference index is not found in the sequence dictionary in the header for this record.
      */
     public void setMateReferenceIndex(final int mateReferenceIndex) {
-        if (mateReferenceIndex == NO_ALIGNMENT_REFERENCE_INDEX) {
-            mMateReferenceIndex = NO_ALIGNMENT_REFERENCE_INDEX;
-            mMateReferenceName = NO_ALIGNMENT_REFERENCE_NAME;
-        }
-        else if (null == mHeader) {
-            throw new IllegalStateException("A non-null SAMFileHeader is required to resolve the mate reference index");
+        // resolveNameFromIndex throws if the index can't be resolved
+        setMateReferenceName(resolveNameFromIndex(mateReferenceIndex, mHeader));
+        mMateReferenceIndex = mateReferenceIndex;
```

Comment:
Same comment as for `setReferenceIndex()` -- Won't setMateReferenceName() set the reference index for us (since we know the header lookup succeeded if resolveNameFromIndex() did not throw)?